curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/6656262ec602ff688c50c8a9.md

```diff
@@ -0,0 +1,165 @@
+---
+id: 6656262ec602ff688c50c8a9
+title: Step 54
+challengeType: 20
+dashedName: step-54
+---
+
+# --description--
+
+After the `if` statement you wrote in the previous step, ensure that `solve()` and `analyze()` methods have not been called yet, otherwise you would keep adding the same items to the `results` and `details` lists.
+
+Check if the `results` and `details` attributes of `equation` contain elements. If not, call the `solve` method, and the `analyze` method, respectively.
```

Comment:
Using class this way is a bit more complicated than it should. Couldn't `solver` function get the _results_ from what `solve` method would return, and _details_ from what `analyze` would return? The strange relation between two is also visible later, when `analyze` is changed to save strings aligned to specifically match `solver`'s formatting.

Both these methods could keep the calculations internally saved, so it would be done just once in any case. Ie. when `solve` would be called again, it would just return the saved result.